cargo-miri/src/phases.rs

```diff
@@ -105,149 +105,148 @@ pub fn phase_cargo_miri(mut args: impl Iterator<Item = String>) {
+        // harder.
+        let cargo_miri_path = env::current_exe()
+            .expect("current executable path invalid")
+            .into_os_string()
+            .into_string()
+            .expect("current executable path is not valid UTF-8");
+        let cargo_cmd = match subcommand {
+            MiriCommand::Forward(s) => s,
+            MiriCommand::Setup => return, // `cargo miri setup` stops here.
+            MiriCommand::Clean => unreachable!(),
+        };
+        let metadata = get_cargo_metadata();
+        let mut cmd = cargo();
+        cmd.arg(&cargo_cmd);
+        // In nextest we have to also forward the main `verb`.
+        if cargo_cmd == "nextest" {
+            cmd.arg(args.next().unwrap_or_else(|| {
+                show_error!("`cargo miri nextest` expects a verb (e.g. `run`)")
+            }));
+        }
+        // We set the following flags *before* forwarding more arguments.
+        // This is needed to fix <https://github.com/rust-lang/miri/issues/2829>: cargo will stop
+        // interpreting things as flags when it sees the first positional argument.
+
+        // Make sure the build target is explicitly set.
+        // This is needed to make the `target.runner` settings do something,
+        // and it later helps us detect which crates are proc-macro/build-script
+        // (host crates) and which crates are needed for the program itself.
         cmd.arg("--target");
-        cmd.arg(host);
-    }
+        cmd.arg(target);
+
+        // Set ourselves as runner for al binaries invoked by cargo.
+        // We use `all()` since `true` is not a thing in cfg-lang, but the empty conjunction is. :)
+        let cargo_miri_path_for_toml = escape_for_toml(&cargo_miri_path);
+        cmd.arg("--config")
+            .arg(format!("target.'cfg(all())'.runner=[{cargo_miri_path_for_toml}, 'runner']"));
+
+        // Set `--target-dir` to `miri` inside the original target directory.
+        let target_dir = get_target_dir(&metadata);
+        cmd.arg("--target-dir").arg(target_dir);
+
+        // Store many-seeds argument.
+        let mut many_seeds = None;
+        // *After* we set all the flags that need setting, forward everything else. Make sure to skip
+        // `--target-dir` (which would otherwise be set twice) and `--many-seeds` (which is our flag, not cargo's).
+        for arg in
+            ArgSplitFlagValue::from_string_iter(&mut args, "--target-dir").filter_map(Result::err)
+        {
+            if arg == "--many-seeds" {
+                many_seeds = Some(DEFAULT_MANY_SEEDS.to_owned());
+            } else if let Some(val) = arg.strip_prefix("--many-seeds=") {
+                many_seeds = Some(val.to_owned());
+            } else {
+                cmd.arg(arg);
+            }
+        }
+        // Forward all further arguments after `--` (not consumed by `ArgSplitFlagValue`) to cargo.
+        cmd.args(args);
```

Comment:
Since this is no longer in the loop the first issue should be addressed.

I added a test that passes 2 targets to miri and expects to see 2 executions of the tests. Is this what you had in mind?

I find the way I did it not so elegant. I am not sure if there is a better way.

If the `cargo_miri` method in the test suite didn't always forward the `--target` flag from its command line to miri, the code might look a bit less complex. But I am not sure if this change is wanted, or if it should be included in this PR.

Reply:
We must always set `--target`, things will break horribly if we invoke cargo without `--target`.

Reply:
Sorry, let me rephrase. We should set `--target`, yes. I wanted to say that the `cargo_miri` function should not always pass the value of `ARGS.target` to pass the `--target` flag. Maybe it would be simpler if `cargo_miri` accepted the targets as an argument.

Then we could specify exactly which target(s) we want, without having to do all the manipulation I do to ensure that exactly two targets are passed to miri in the tests.

Perhaps we can add a `targets` argument with a default None value. If the value is none, only then we try to use `ARGS.target`, otherwise we use the explicitly provided argument.

Reply:
Ah, you are talking about run_test.py, not cargo-miri. That context was missing. :)

Yes, please make cargo_miri take a "targets" parameter that defaults to `None`, in which case it uses ARGS.target.

Reply:
yes, i was talking about `run_test.py`.

I changed the code as discussed. Let me know if that works!